Add episode 2 on MMSF theory (#7)
esciencecenter-digital-skills/model-coupling-workshop: episodes/02-theory.md

vary throughout the domain, pick the smallest. The *extent* refers to the size
of the process in space and time, so the size of its domain. How long does it
take to complete (or to reach a steady state), and how large an area is
modelled?

::::::::::::::::::::::::::::::::::::: instructor
- Add to shared document
::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: challenge

## Challenge 2: Models and scales

What are the scales of the models you previously considered the domains of? What
are their grain and extent in space and in time?

:::::::::::::::::::::::: solution

## Example solution

### Surface

Model 1 simulates the heating of the Earth's surface as the sun shines on it.

The temperature of the Earth's surface only changes slowly during the day, and
we can probably say there are no significant changes from one minute to the
next, or even for a somewhat larger interval. So the time grain of model 1 is on
the order of a few minutes, perhaps up to an hour. The time extent of the model
is one day, as it will repeat itself after that.

In space, the grain depends on the research question, and could be as small as

> 10cm if we are looking at the detailed thermal environment around a building, or

Reviewer: ```suggestion
10 cm if we are looking at the detailed thermal environment around a building, or
```

it.

As you can see, we can plot multiple models on a Scale Separation Map and when
we do, the map shows the relationship between the scales of two models. Both
horizontally and vertically, model rectangles can have a gap between them (i.e.
be separated), or be adjacent, or overlap, and the corresponding models are said
to be scale separated, scale adjacent or scale overlapping in space and/or in
time.

For coupling, it is actually only these relationships that matter. Even if you
don't know the exact grain or extent of a particular model, you can often decide
whether it's smaller, larger or the same as the grain or extent of another model
and draw them correspondingly.

::::::::::::::::::::::::::::::::::::: instructor
- All together in class
::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: challenge

## Challenge 3: Scale Separation Map

- What are the relations of scales a through h in the figure relative to the
  reference scale?
- Draw a scale separation map for the models you would like to couple.

:::::::::::::::::::::::: solution

## Example solution

### Scale relations

a. Spatially (larger) and temporally (faster) scale separated.

> b. Spatially scale adjacent (larger), tomporally scale overlapping.

Reviewer: ```suggestion
b. Spatially scale adjacent (larger), temporally scale overlapping.
```

The three cases above demonstrate the four *Coupling Templates* defined by the
MMSF. The first one, $O_f$ to $f_{init}$, is called *dispatch*. The second one,
$O_i$ to $S$ is called *interact*, and usually comes in pairs. The third one
and the fourth one usually go together, as the combination *call* ($O_i$ to
$f_{init}$) and *release* ($O_f$ to $S$).

Given the constraints of which operators can send and which can receive, these
are in fact all four possible types of connections, and between them they cover
all kinds of temporal scale relations.

If the time domains or time scales overlap, but are not equivalent, then an
additional component is needed that sits between the models; we will not go
into that advanced use case here.

## Spatial domains, scales and multiplicity

::::::::::::::::::::::::::::::::::::: instructor
- Space is also important of course
- Again three scenarios
- --- **#** Same domain and scale
- --- --- **#** May be best to make a single program
- --- **#** Same spatial domain, adjacent or separated scales
- --- --- **#** Multiple instances of the small one!
- --- **#** Adjacent spatial domains, same or overlapping spatial scales

> - --- --- **#** For time-scale separation, fast model needs to keep state!

Reviewer: ```suggestion
- --- --- **#** For time-scale separation, the fast model needs to keep state!
```
